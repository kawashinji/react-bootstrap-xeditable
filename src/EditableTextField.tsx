import React, { ChangeEvent } from 'react';
import XEditable from './XEditable';

export interface IEditableTextFieldProps extends React.InputHTMLAttributes<HTMLInputElement> {
  isEditing?: boolean;
  isEditable?: boolean;
  linkClassName?: string;
  nonValueText?: string;
  value: string;
  onSave: (value: string | number | string[]) => void;
  onCancel?: () => void;
}

interface IState {
  isEditing: boolean;
  value: string | number | string[];
}

export default class EditableTextField extends React.Component<IEditableTextFieldProps, IState> {
  static defaultProps: Partial<IEditableTextFieldProps> = {
    isEditable: true
  };

  state: IState = {
    isEditing: !!this.props.isEditing,
    value: this.props.value || ''
  };

  onChange = (event: ChangeEvent<HTMLInputElement>) => {
    event.preventDefault();
    this.setState({
      value: event.currentTarget.value
    });

    if (this.props.onChange) {
      this.props.onChange(event);
    }
  };

  onSave = () => {
    this.setState({
      isEditing: false
    });
    this.props.onSave(this.state.value);
  };

  onCancel = () => {
    this.setState({
      isEditing: false,
      value: this.props.value || ''
    });

    if (this.props.onCancel) {
      this.props.onCancel();
    }
  };

  onClear = () => {
    this.setState({
      value: ''
    });
  };

  onLinkClick = () => {
    if (this.props.isEditable) {
      this.setState({
        isEditing: true
      });
    }
  };

  renderEditing() {
    const { isEditing, isEditable, onSave, onCancel, ...otherProps } = this.props;
    return (
      <XEditable onSave={this.onSave} onCancel={this.onCancel}>
        <input {...otherProps} onChange={this.onChange} value={this.state.value} />
        <span className="editable-clear-x" onClick={this.onClear} />
      </XEditable>
    );
  }

  renderIcon() {
    if (!this.state.value || this.state.value === '') {
      return <span className="glyphicon glyphicon-pencil" />;
    }
    return;
  }

  render() {
    if (!this.props.isEditable) {
      return <div>{this.state.value || this.props.nonValueText}</div>;
    }
    const classNames = ['editable', 'editable-click'];

    if (!this.state.value) {
      classNames.push('editable-empty');
    }

    if (this.props.linkClassName) {
      classNames.concat(this.props.linkClassName.split(' '));
    }

    if (this.state.isEditing) {
      return this.renderEditing();
    } else {
      return (
        <a href="javascript:;" className={classNames.join(' ')} onClick={this.onLinkClick}>
          {this.props.value || this.props.nonValueText}
          {this.renderIcon()}
        </a>
      );
    }
  }
}
