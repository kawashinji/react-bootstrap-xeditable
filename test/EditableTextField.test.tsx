import 'jest';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import React from 'react';
import renderer from 'react-test-renderer';
import EditableTextField from '../src/EditableTextField';

it('renders correctly', () => {
  const tree = renderer.create(<EditableTextField name="test" value="Hello world" onSave={() => {}} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('Render EditableTextField and Click it', () => {
  const wrapper = mount(<EditableTextField name="test" value="Hello world" onSave={() => {}} />);
  wrapper.find('a').simulate('click');
  expect(toJSON(wrapper)).toMatchSnapshot();
});
