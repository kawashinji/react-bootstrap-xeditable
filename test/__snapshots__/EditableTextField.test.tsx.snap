// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render EditableTextField and Click it 1`] = `
<EditableTextField
  isEditable={true}
  name="test"
  onSave={[Function]}
  value="Hello world"
>
  <XEditable
    isLoading={false}
    onCancel={[Function]}
    onSave={[Function]}
  >
    <span
      className="editable-container editable-inline"
    >
      <div
        className="form-inline editableform"
      >
        <div
          className="control-group form-group"
        >
          <div>
            <div
              className="editable-input"
              style={
                Object {
                  "position": "relative",
                }
              }
            >
              <input
                name="test"
                onChange={[Function]}
                value="Hello world"
              />
              <span
                className="editable-clear-x"
                onClick={[Function]}
              />
            </div>
            <div
              className="editable-buttons"
            >
              <button
                className="btn btn-primary btn-sm editable-submit"
                onClick={[Function]}
                type="submit"
              >
                <i
                  className="glyphicon glyphicon-ok"
                />
              </button>
              <button
                className="btn btn-default btn-sm editable-cancel"
                onClick={[Function]}
                type="button"
              >
                <i
                  className="glyphicon glyphicon-remove"
                />
              </button>
            </div>
          </div>
          <div
            className="editable-error-block help-block"
          />
        </div>
      </div>
    </span>
  </XEditable>
</EditableTextField>
`;

exports[`renders correctly 1`] = `
<a
  className="editable editable-click"
  href="javascript:;"
  onClick={[Function]}
>
  Hello world
</a>
`;
